core/trino-main/src/main/java/io/trino/likematcher/FjsMatcher.java

```diff
@@ -0,0 +1,210 @@
+
+                if (element instanceof Pattern.Literal literal) {
+                    checkArgument(i == 0 || !(pattern.get(i - 1) instanceof Pattern.Literal), "Multiple consecutive literals found");
+                    byte[] bytes = literal.value().getBytes(StandardCharsets.UTF_8);
+                    patterns.add(bytes);
+                    bmsShifts.add(computeBmsShifts(bytes));
+                    kmpShifts.add(computeKmpShifts(bytes));
+                }
+                else if (element instanceof Pattern.Any) {
+                    throw new IllegalArgumentException("'any' pattern not supported");
+                }
+            }
+        }
+
+        private static int[] computeKmpShifts(byte[] pattern)
+        {
+            int[] result = new int[pattern.length + 1];
+            result[0] = -1;
+
+            int j = -1;
+            for (int i = 1; i < result.length; i++) {
+                while (j >= 0 && pattern[i - 1] != pattern[j]) {
+                    j = result[j];
+                }
+                j++;
+                result[i] = j;
+            }
+
+            return result;
+        }
+
+        private static int[] computeBmsShifts(byte[] pattern)
+        {
+            int[] result = new int[256];
+
+            for (int i = 0; i < pattern.length; i++) {
+                result[pattern[i] & 0xFF] = i + 1;
+            }
+
+            return result;
+        }
+
+        private static int find(byte[] input, final int offset, final int length, byte[] pattern, int[] bmsShifts, int[] kmpShifts)
+        {
+            if (pattern.length > length || pattern.length == 0) {
+                return -1;
+            }
+
+            final int inputLimit = offset + length;
+
+            int i = offset;
+            while (true) {
+                // Attempt to match the last position of the pattern
+                // As long as it doesn't match, skip ahead based on the Boyer-Moore-Sunday heuristic
+                int matchEnd = i + pattern.length - 1;
+                while (matchEnd < inputLimit - 1 && input[matchEnd] != pattern[pattern.length - 1]) {
+                    int shift = pattern.length + 1 - bmsShifts[input[matchEnd + 1] & 0xFF];
+                    matchEnd += shift;
+                }
+
+                if (matchEnd == inputLimit - 1 && match(input, inputLimit - pattern.length, pattern)) {
+                    return inputLimit - pattern.length;
+                }
+                else if (matchEnd >= inputLimit - 1) {
+                    return -1;
+                }
+
+                // At this point, we know the last position of the pattern matches with some
+                // position in the input text given by "matchEnd"
+                // Use KMP to match the first length-1 characters of the pattern
+
+                i = matchEnd - (pattern.length - 1);
+
+                int j = findLongestMatch(input, i, pattern, 0, pattern.length - 1);
```

Comment:
wondering if we can have ore descriptive names here for variables. it's a bit tricky because meaning keeps somewhat changing

i => currentInputOffset
j => currentPatternPrefixMatchLength

Reply:
I went with the prevalent convention used when people write string matching algorithms. Commonly:
* `n`: length of the input string
* `m`: length of the pattern
* `i`: index into the input stream
* `j`: index into the pattern.